Add trusted funding to bos open and inbound-channel-rules

niteshbalusu11 opened the pull request:
Added trusted funding to bos open and to inbound_channel_rules for interception.
Updated ln-sync to latest version.

Example bos command
```
bos open --node alice 029b107dfd3395723d910361abf3a227ca7d6f01dd21d7d1f8ca16bf749aa86fad --amount 100000 --trusted=true 035500596dc846f7e951bb4563533abb2cfa37c890f43790c43e7ec5731f751fb7 --amount 100000 --trusted=false
```

Example inbound-channel-rules command:
```
bos inbound-channel-rules --node carol --rule "EXACT(public_key, '03fcf275c617d9b9e7a5dcde2d1c77f469fdbe844cabefec76db62c526acbc662d')" --rule "is_trusted_funding"
```

alexbosworth commented:
seems like it needs to be rebased on master

niteshbalusu11 commented:
I think it’s because I included the bumped version of ln-sync and you had already committed that yesterday.

alexbosworth commented on bos:
```diff
@@ -1131,6 +1134,7 @@ prog
   .option('--node <node_name>', 'Saved node to open channels')
   .option('--opening-node <node_name>', 'Open with saved node', REPEATABLE)
   .option('--set-fee-rate <ppm>', 'Set forward fee rate to peer', REPEATABLE)
+  .option('--trusted <is_trusted_funding>', 'Open a trusted channel (true/false)', REPEATABLE)
```
maybe type should be expanded instead of adding a new flag? like pubilc-trusted and private-trusted?

niteshbalusu11 replied:
You mean the type flag should have 4 options?

alexbosworth replied:
Yeah I think 2 more options is ok

niteshbalusu11 replied:
Cool. Ok.

alexbosworth commented on peers/open_channels.js:
```diff
@@ -186,8 +202,42 @@ module.exports = (args, cbk) => {
         return cbk(null, capacities);
       }],
 
+      // Get the wallet version to make sure the node supports internal funding
```
Is there a reason version checking is necessary? I think it can be skipped

niteshbalusu11 replied:
Just added it because it's unsupported in lower versions. We should just let it fail on its own?

alexbosworth replied:
Yeah just let it fail, version checking isn't very reliable because people compile themselves

niteshbalusu11 replied:
Cool ok.

alexbosworth commented on peers/open_channels.js:
```diff
@@ -745,15 +831,14 @@ module.exports = (args, cbk) => {
 
             // Every channel to open should be reflected in a pending channel
-            if (opening.length !== toOpen.length) {
+            if ((opening.length + getTrustedChannels.length) !== toOpen.length) {
```
This check won't work if the trusted channel takes time to show up in the channels list, to safeguard against it taking time channels should be pulled in the asyncMap above along with getPending channels. It doesn't need to look if they are trusted channels though, just that the transaction id is matching

niteshbalusu11 replied:
I added an async retry above while getting trusted channels from all channels list and that will only pass if it gets all trusted channels and will only enter this part after that.

alexbosworth replied:
Ah cool ok, it still feels like overkill though

I'd just do a single call that combines getPending and getChannels where getPending is right now and then use the same logic across both result sets

alexbosworth commented:
I'm kind of feeling like only rules is maybe not enough for trusted funding, as an extra safety check I think inbound channel rules should require a whitelist of public keys where trusted funding is allowed

alexbosworth commented on peers/open_request_violation.js:
```diff
@@ -58,6 +63,7 @@ module.exports = args => {
     capacity: args.capacity,
     channel_ages: args.channel_ages,
     fee_rates: args.fee_rates,
+    is_trusted_funding: args.is_trusted_funding,
```
how about just `trusted_funding` rather than `is_trusted_funding`

niteshbalusu11 replied:
Yeah. I'll make that change.

niteshbalusu11 commented:
> I'm kind of feeling like only rules is maybe not enough for trusted funding, as an extra safety check I think inbound channel rules should require a whitelist of public keys where trusted funding is allowed

May be a separate flag to take pubkeys? Stuffing multiple pubkeys in the rule makes it really hard to type it with OR conditions.

alexbosworth commented:
Yeah a separate flag like `bos inbound-channel-rules --allow-trusted-funding-from KEY --rule "capacity > 5*m"`?

niteshbalusu11 commented:
Cool yeah.

alexbosworth commented:
If the nodes are on the whitelist also maybe that's easy because also you don't need to specify trusted funding key on every rule so the workflow for trusted channels is

1. Both sides update conf file to include trusted channels support
2. Receiving side runs `bos inbound-channel-rules --trust-funding-from KEY`
3. Proposing side does `bos open --type private-trusted KEY`

niteshbalusu11 commented:
Ok. Will make the changes.

niteshbalusu11 commented:
Should there be an option, if all channels you're opening are trusted, you give a choice to never broadcast the tx? Like give an option to a have a channel but never settle.

alexbosworth commented:
Sounds like a nice feature, could be a new option on open like `--avoid-broadcast` or something

niteshbalusu11 commented:
If a such a channel ever has to be closed, you basically just have to abandon it?

alexbosworth commented:
I'm not sure how close works, I didn't play with that

niteshbalusu11 commented:
I updated the trusted funding flow like this:

Both sides update conf file to include trusted channels support
Receiving side runs bos inbound-channel-rules --trust-funding-from KEY
Proposing side does bos open --type private-trusted KEY

and also added --avoid-broadcast.

The only thing I couldn't figure out is how to merge the task of getting pendingchannels and trusted channels in one task before broadcasting.

niteshbalusu11 commented on peers/open_channels.js:
```diff
@@ -186,6 +173,16 @@ module.exports = (args, cbk) => {
 
+        if (!!args.avoid_broadcast) {
+          if (!args.types.length) {
+            return cbk([400, 'ExpectedTypesToAvoidBroadcast']);
```
The next line just checks if all types should be public-trusted or private-trusted so added to check types need to exist to avoid broadcast.

niteshbalusu11 commented on peers/open_channels.js:
```diff
@@ -186,6 +173,16 @@ module.exports = (args, cbk) => {
+          }
+
+          if ((args.types.filter(n => !!trustedFundingTypes.includes(n))).length !== args.types.length) {
+            return cbk([400, 'AllTypesMustBeTrustedFundingToAvoidBroadcast'])
```
Cool ok. I can skip the check.

alexbosworth commented on peers/open_channels.js:
```diff
@@ -817,6 +784,16 @@ module.exports = (args, cbk) => {
           return cbk();
         }
 
+        // Exit early if avoiding broadcast
```
This should be moved down just before the broadcast because it wouldn't be safe for the caller to self-publish this transaction yet

niteshbalusu11 commented:
Fixed them.

niteshbalusu11 commented:
There is one issue I noticed, if you try to set the fee rates, it does not return you the broadcasted the tx ID until the fee rates have been applied to the channels pending open. Maybe there needs to be a logger after the broadcast happens. 

Also, if we are doing --avoid-broadcast can you return the raw tx instead of logging it? It'll help me with LndBoss.

alexbosworth commented:
OK I'll look at that
